# Stock Sentiment Analysis using NLP

This project intends to predict the movement of stock price by conducting sentiment analysis on news headlines. The dataset used is from https://www.kaggle.com/aaron7sun/stocknews.

### Installing Dependencies

In [9]:
pip install vaderSentiment textblob

Note: you may need to restart the kernel to use updated packages.


### Loading Required Libraries

In [10]:
import pandas as pd
import numpy as np
from textblob import TextBlob
import re
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis

### Loading Datasets

In [26]:
news = pd.read_csv("Combined_News_DJIA.csv")
prices = pd.read_csv("upload_DJIA_table.csv")

In [27]:
news.head(3)

Date  Label                                               Top1  \
0  2008-08-08      0  b"Georgia 'downs two Russian warplanes' as cou...   
1  2008-08-11      1  b'Why wont America and Nato help us? If they w...   
2  2008-08-12      0  b'Remember that adorable 9-year-old who sang a...   

                                          Top2  \
0      b'BREAKING: Musharraf to be impeached.'   
1  b'Bush puts foot down on Georgian conflict'   
2           b"Russia 'ends Georgia operation'"   

                                                Top3  \
0  b'Russia Today: Columns of troops roll into So...   
1  b"Jewish Georgian minister: Thanks to Israeli ...   
2  b'"If we had no sexual harassment we would hav...   

                                                Top4  \
0  b'Russian tanks are moving towards the capital...   
1  b'Georgian army flees in disarray as Russians ...   
2  b"Al-Qa'eda is losing support in Iraq because ...   

                                                Top5  \
0  b"Afghan children raped with 'impunity,' U.N. ...   
1      b"Olympic opening ceremony fireworks 'faked'"   
2  b'Ceasefire in Georgia: Putin Outmaneuvers the...   

                                                Top6  \
0  b'150 Russian tanks have entered South Ossetia...   
1  b'What were the Mossad with fraudulent New Zea...   
2  b'Why Microsoft and Intel tried to kill the XO...   

                                                Top7  \
0  b"Breaking: Georgia invades South Ossetia, Rus...   
1  b'Russia angered by Israeli military sale to G...   
2  b'Stratfor: The Russo-Georgian War and the Bal...   

                                                Top8  ...  \
0  b"The 'enemy combatent' trials are nothing but...  ...   
1  b'An American citizen living in S.Ossetia blam...  ...   
2  b"I'm Trying to Get a Sense of This Whole Geor...  ...   

                                               Top16  \
0  b'Georgia Invades South Ossetia - if Russia ge...   
1  b'Israel and the US behind the Georgian aggres...   
2  b'U.S. troops still in Georgia (did you know t...   

                                               Top17  \
0                b'Al-Qaeda Faces Islamist Backlash'   
1  b'"Do not believe TV, neither Russian nor Geor...   
2       b'Why Russias response to Georgia was right'   

                                               Top18  \
0  b'Condoleezza Rice: "The US would not act to p...   
1  b'Riots are still going on in Montreal (Canada...   
2  b'Gorbachev accuses U.S. of making a "serious ...   

                                               Top19  \
0  b'This is a busy day:  The European Union has ...   
1    b'China to overtake US as largest manufacturer'   
2         b'Russia, Georgia, and NATO: Cold War Two'   

                                               Top20  \
0  b"Georgia will withdraw 1,000 soldiers from Ir...   
1                     b'War in South Ossetia [PICS]'   
2  b'Remember that adorable 62-year-old who led y...   

                                               Top21  \
0  b'Why the Pentagon Thinks Attacking Iran is a ...   
1  b'Israeli Physicians Group Condemns State Tort...   
2          b'War in Georgia: The Israeli connection'   

                                               Top22  \
0  b'Caucasus in crisis: Georgia invades South Os...   
1  b' Russia has just beaten the United States ov...   
2  b'All signs point to the US encouraging Georgi...   

                                               Top23  \
0  b'Indian shoe manufactory  - And again in a se...   
1  b'Perhaps *the* question about the Georgia - R...   
2  b'Christopher King argues that the US and NATO...   

                                               Top24  \
0  b'Visitors Suffering from Mental Illnesses Ban...   
1                 b'Russia is so much better at war'   
2                        b'America: The New Mexico?'   

                                               Top25  
0           b"No Help for Mexico's Kidnapping Surge"  
1 

In [25]:
prices.head(3)

Date          Open          High           Low         Close  \
0  2016-07-01  17924.240234  18002.380859  17916.910156  17949.369141   
1  2016-06-30  17712.759766  17930.609375  17711.800781  17929.990234   
2  2016-06-29  17456.019531  17704.509766  17456.019531  17694.679688   

      Volume     Adj Close  
0   82160000  17949.369141  
1  133030000  17929.990234  
2  106380000  17694.679688

### Data Pre-Processing

In [28]:
#Merging the headlines and prices dataframes
data = news.merge(prices, how="inner", on="Date", left_index=True)

In [29]:
data.head(3)

Date  Label                                               Top1  \
1988  2008-08-08      0  b"Georgia 'downs two Russian warplanes' as cou...   
1987  2008-08-11      1  b'Why wont America and Nato help us? If they w...   
1986  2008-08-12      0  b'Remember that adorable 9-year-old who sang a...   

                                             Top2  \
1988      b'BREAKING: Musharraf to be impeached.'   
1987  b'Bush puts foot down on Georgian conflict'   
1986           b"Russia 'ends Georgia operation'"   

                                                   Top3  \
1988  b'Russia Today: Columns of troops roll into So...   
1987  b"Jewish Georgian minister: Thanks to Israeli ...   
1986  b'"If we had no sexual harassment we would hav...   

                                                   Top4  \
1988  b'Russian tanks are moving towards the capital...   
1987  b'Georgian army flees in disarray as Russians ...   
1986  b"Al-Qa'eda is losing support in Iraq because ...   

                                                   Top5  \
1988  b"Afghan children raped with 'impunity,' U.N. ...   
1987      b"Olympic opening ceremony fireworks 'faked'"   
1986  b'Ceasefire in Georgia: Putin Outmaneuvers the...   

                                                   Top6  \
1988  b'150 Russian tanks have entered South Ossetia...   
1987  b'What were the Mossad with fraudulent New Zea...   
1986  b'Why Microsoft and Intel tried to kill the XO...   

                                                   Top7  \
1988  b"Breaking: Georgia invades South Ossetia, Rus...   
1987  b'Russia angered by Israeli military sale to G...   
1986  b'Stratfor: The Russo-Georgian War and the Bal...   

                                                   Top8  ...  \
1988  b"The 'enemy combatent' trials are nothing but...  ...   
1987  b'An American citizen living in S.Ossetia blam...  ...   
1986  b"I'm Trying to Get a Sense of This Whole Geor...  ...   

                                                  Top22  \
1988  b'Caucasus in crisis: Georgia invades South Os...   
1987  b' Russia has just beaten the United States ov...   
1986  b'All signs point to the US encouraging Georgi...   

                                                  Top23  \
1988  b'Indian shoe manufactory  - And again in a se...   
1987  b'Perhaps *the* question about the Georgia - R...   
1986  b'Christopher King argues that the US and NATO...   

                                                  Top24  \
1988  b'Visitors Suffering from Mental Illnesses Ban...   
1987                 b'Russia is so much better at war'   
1986                        b'America: The New Mexico?'   

                                                  Top25          Open  \
1988           b"No Help for Mexico's Kidnapping Surge"  11432.089844   
1987  b"So this is what it's come to: trading sex fo...  11729.669922   
1986  b"BBC NEWS | Asia-Pacific | Extinction 'by man...  11781.700195   

              High           Low         Close     Volume     Adj Close  
1988  11759.959961  11388.040039  11734.320312  212830000  11734.320312  
1987  11867.110352  11675.530273  11782.349609  183190000  11782.349609  
1986  11782.349609  11601.519531  11642.469727  173590000  11642.469727  

[3 rows x 33 columns]

In [30]:
#Combining the Top 25 headlines for each row
headlines = []

for row in range(0, len(data.index)):
    headlines.append(' '.join( str(x) for x in data.iloc[row, 2:27] ))

In [31]:
headlines[0]

'b"Georgia \'downs two Russian warplanes\' as countries move to brink of war" b\'BREAKING: Musharraf to be impeached.\' b\'Russia Today: Columns of troops roll into South Ossetia; footage from fighting (YouTube)\' b\'Russian tanks are moving towards the capital of South Ossetia, which has reportedly been completely destroyed by Georgian artillery fire\' b"Afghan children raped with \'impunity,\' U.N. official says - this is sick, a three year old was raped and they do nothing" b\'150 Russian tanks have entered South Ossetia whilst Georgia shoots down two Russian jets.\' b"Breaking: Georgia invades South Ossetia, Russia warned it would intervene on SO\'s side" b"The \'enemy combatent\' trials are nothing but a sham: Salim Haman has been sentenced to 5 1/2 years, but will be kept longer anyway just because they feel like it." b\'Georgian troops retreat from S. Osettain capital, presumably leaving several hundred people killed. [VIDEO]\' b\'Did the U.S. Prep Georgia for War with Russia?\'

In [35]:
#Cleaning the headlines array by removing unwanted characters
clean_headlines = []

for i in range(0, len(headlines)):
    clean_headlines.append(re.sub("b[(')]", '', headlines[i])) #removing occurrences of b' and replacing it with empty string
    clean_headlines[i] = re.sub('b[(")]', '', clean_headlines[i]) #removing occurrences of b" and replacing it with empty string
    clean_headlines[i] = re.sub("\'", '', clean_headlines[i]) #removing occurrences of \' and replacing it with empty string

In [38]:
clean_headlines[0]

'Georgia downs two Russian warplanes as countries move to brink of war" BREAKING: Musharraf to be impeached. Russia Today: Columns of troops roll into South Ossetia; footage from fighting (YouTube) Russian tanks are moving towards the capital of South Ossetia, which has reportedly been completely destroyed by Georgian artillery fire Afghan children raped with impunity, U.N. official says - this is sick, a three year old was raped and they do nothing" 150 Russian tanks have entered South Ossetia whilst Georgia shoots down two Russian jets. Breaking: Georgia invades South Ossetia, Russia warned it would intervene on SOs side" The enemy combatent trials are nothing but a sham: Salim Haman has been sentenced to 5 1/2 years, but will be kept longer anyway just because they feel like it." Georgian troops retreat from S. Osettain capital, presumably leaving several hundred people killed. [VIDEO] Did the U.S. Prep Georgia for War with Russia? Rice Gives Green Light for Israel to Attack Iran: S

In [39]:
#Add back clean_headlines to our data dataframe
data["ConcatenatedTopNews"] = clean_headlines

In [40]:
data.head(3)

Date  Label                                               Top1  \
1988  2008-08-08      0  b"Georgia 'downs two Russian warplanes' as cou...   
1987  2008-08-11      1  b'Why wont America and Nato help us? If they w...   
1986  2008-08-12      0  b'Remember that adorable 9-year-old who sang a...   

                                             Top2  \
1988      b'BREAKING: Musharraf to be impeached.'   
1987  b'Bush puts foot down on Georgian conflict'   
1986           b"Russia 'ends Georgia operation'"   

                                                   Top3  \
1988  b'Russia Today: Columns of troops roll into So...   
1987  b"Jewish Georgian minister: Thanks to Israeli ...   
1986  b'"If we had no sexual harassment we would hav...   

                                                   Top4  \
1988  b'Russian tanks are moving towards the capital...   
1987  b'Georgian army flees in disarray as Russians ...   
1986  b"Al-Qa'eda is losing support in Iraq because ...   

                                                   Top5  \
1988  b"Afghan children raped with 'impunity,' U.N. ...   
1987      b"Olympic opening ceremony fireworks 'faked'"   
1986  b'Ceasefire in Georgia: Putin Outmaneuvers the...   

                                                   Top6  \
1988  b'150 Russian tanks have entered South Ossetia...   
1987  b'What were the Mossad with fraudulent New Zea...   
1986  b'Why Microsoft and Intel tried to kill the XO...   

                                                   Top7  \
1988  b"Breaking: Georgia invades South Ossetia, Rus...   
1987  b'Russia angered by Israeli military sale to G...   
1986  b'Stratfor: The Russo-Georgian War and the Bal...   

                                                   Top8  ...  \
1988  b"The 'enemy combatent' trials are nothing but...  ...   
1987  b'An American citizen living in S.Ossetia blam...  ...   
1986  b"I'm Trying to Get a Sense of This Whole Geor...  ...   

                                                  Top23  \
1988  b'Indian shoe manufactory  - And again in a se...   
1987  b'Perhaps *the* question about the Georgia - R...   
1986  b'Christopher King argues that the US and NATO...   

                                                  Top24  \
1988  b'Visitors Suffering from Mental Illnesses Ban...   
1987                 b'Russia is so much better at war'   
1986                        b'America: The New Mexico?'   

                                                  Top25          Open  \
1988           b"No Help for Mexico's Kidnapping Surge"  11432.089844   
1987  b"So this is what it's come to: trading sex fo...  11729.669922   
1986  b"BBC NEWS | Asia-Pacific | Extinction 'by man...  11781.700195   

              High           Low         Close     Volume     Adj Close  \
1988  11759.959961  11388.040039  11734.320312  212830000  11734.320312   
1987  11867.110352  11675.530273  11782.349609  183190000  11782.349609   
1986  11782.349609  11601.519531  11642.469727  173590000  11642.469727   

                                    ConcatenatedTopNews  
1988  Georgia downs two Russian warplanes as countri...  
1987  Why wont America and Nato help us? If they won...  
1986  Remember that adorable 9-year-old who sang at ...  

[3 rows x 34 columns]

In [43]:
data["ConcatenatedTopNews"][1988]

'Georgia downs two Russian warplanes as countries move to brink of war" BREAKING: Musharraf to be impeached. Russia Today: Columns of troops roll into South Ossetia; footage from fighting (YouTube) Russian tanks are moving towards the capital of South Ossetia, which has reportedly been completely destroyed by Georgian artillery fire Afghan children raped with impunity, U.N. official says - this is sick, a three year old was raped and they do nothing" 150 Russian tanks have entered South Ossetia whilst Georgia shoots down two Russian jets. Breaking: Georgia invades South Ossetia, Russia warned it would intervene on SOs side" The enemy combatent trials are nothing but a sham: Salim Haman has been sentenced to 5 1/2 years, but will be kept longer anyway just because they feel like it." Georgian troops retreat from S. Osettain capital, presumably leaving several hundred people killed. [VIDEO] Did the U.S. Prep Georgia for War with Russia? Rice Gives Green Light for Israel to Attack Iran: S

### Exploratory Data Analysis

In [44]:
#Functions to obtain the polarity and subjectivity of the headlines
def getPolarity(string):
    return TextBlob(string).sentiment.polarity

def getSubjectivity(string):
    return TextBlob(string).sentiment.subjectivity

data["Polarity"] = data["ConcatenatedTopNews"].apply(getPolarity)
data["Subjectivity"] = data["ConcatenatedTopNews"].apply(getSubjectivity)

In [45]:
data.head(3)

Date  Label                                               Top1  \
1988  2008-08-08      0  b"Georgia 'downs two Russian warplanes' as cou...   
1987  2008-08-11      1  b'Why wont America and Nato help us? If they w...   
1986  2008-08-12      0  b'Remember that adorable 9-year-old who sang a...   

                                             Top2  \
1988      b'BREAKING: Musharraf to be impeached.'   
1987  b'Bush puts foot down on Georgian conflict'   
1986           b"Russia 'ends Georgia operation'"   

                                                   Top3  \
1988  b'Russia Today: Columns of troops roll into So...   
1987  b"Jewish Georgian minister: Thanks to Israeli ...   
1986  b'"If we had no sexual harassment we would hav...   

                                                   Top4  \
1988  b'Russian tanks are moving towards the capital...   
1987  b'Georgian army flees in disarray as Russians ...   
1986  b"Al-Qa'eda is losing support in Iraq because ...   

                                                   Top5  \
1988  b"Afghan children raped with 'impunity,' U.N. ...   
1987      b"Olympic opening ceremony fireworks 'faked'"   
1986  b'Ceasefire in Georgia: Putin Outmaneuvers the...   

                                                   Top6  \
1988  b'150 Russian tanks have entered South Ossetia...   
1987  b'What were the Mossad with fraudulent New Zea...   
1986  b'Why Microsoft and Intel tried to kill the XO...   

                                                   Top7  \
1988  b"Breaking: Georgia invades South Ossetia, Rus...   
1987  b'Russia angered by Israeli military sale to G...   
1986  b'Stratfor: The Russo-Georgian War and the Bal...   

                                                   Top8  ...  \
1988  b"The 'enemy combatent' trials are nothing but...  ...   
1987  b'An American citizen living in S.Ossetia blam...  ...   
1986  b"I'm Trying to Get a Sense of This Whole Geor...  ...   

                                                  Top25          Open  \
1988           b"No Help for Mexico's Kidnapping Surge"  11432.089844   
1987  b"So this is what it's come to: trading sex fo...  11729.669922   
1986  b"BBC NEWS | Asia-Pacific | Extinction 'by man...  11781.700195   

              High           Low         Close     Volume     Adj Close  \
1988  11759.959961  11388.040039  11734.320312  212830000  11734.320312   
1987  11867.110352  11675.530273  11782.349609  183190000  11782.349609   
1986  11782.349609  11601.519531  11642.469727  173590000  11642.469727   

                                    ConcatenatedTopNews  Polarity Subjectivity  
1988  Georgia downs two Russian warplanes as countri... -0.048568     0.267549  
1987  Why wont America and Nato help us? If they won...  0.121956     0.374806  
1986  Remember that adorable 9-year-old who sang at ... -0.044302     0.536234  

[3 rows x 36 columns]

Polarity is a float ranges between -1.0 and 1.0 where -1.0 indicates negative and 1.0 indicates positive. Subjectivity is a float ranges between 0.0 and 1.0 where 0.0 indicates objective and 1.0 indicates subjective. Let us now obtain the sentiment scores of the headlines.

In [48]:
#Function to obtain the sentiment scores of the headlines
def getSentimentScores(string):
    sia = SentimentIntensityAnalyzer()
    sentiment = sia.polarity_scores(string)
    return sentiment

In [49]:
#Obtaining the sentiment scores for each day
compound = []
negative = []
positive = []
neutral = []
SIA = 0

for i in range(0, len(data["ConcatenatedTopNews"])):
    SIA = getSentimentScores(data["ConcatenatedTopNews"][i])
    compound.append(SIA['compound'])
    negative.append(SIA['neg'])
    positive.append(SIA['pos'])
    neutral.append(SIA['neu'])

**VADER**

VADER stands for Valence Aware Dictionary for sEntiment Reasoning. VADER is a less resource-consuming sentiment analysis model that uses a set of rules to specify a mathematical model without explicitly coding it. VADER consumes fewer resources as compared to Machine Learning models as there is no need for vast amounts of training data. VADER’s resource-efficient approach helps us to decode and quantify the emotions contained in streaming media such as text, audio or video. VADER doesn’t suffer severely from a speed-performance tradeoff.

**Valence Score**

It is a score assigned to the word under consideration by means of observation and experiences rather than pure logic. Valence score is measured on a scale from -4 to +4, where -4 stands for the most ‘Negative’ sentiment and +4 for the most ‘Positive’ sentiment. The midpoint 0 represents ‘Neutral’ Sentiment.

**Compound Scores**

The compound score is computed by summing the valence scores of each word in the lexicon, adjusted according to the rules, and then normalized to be between -1 (most extreme negative) and +1 (most extreme positive). This is the most useful metric if you want a single unidimensional measure of sentiment for a given sentence.

$$x = \frac{x}{\sqrt{x^2 + α}}$$

where x = sum of valence scores of constituent words, and α = Normalization constant (default value is 15).

In [50]:
#Store scores in the data dataframe
data["Compound"] = compound
data["Negative"] = negative
data["Positive"] = positive
data["Neutral"] = neutral

In [51]:
data.head(3)

Date  Label                                               Top1  \
1988  2008-08-08      0  b"Georgia 'downs two Russian warplanes' as cou...   
1987  2008-08-11      1  b'Why wont America and Nato help us? If they w...   
1986  2008-08-12      0  b'Remember that adorable 9-year-old who sang a...   

                                             Top2  \
1988      b'BREAKING: Musharraf to be impeached.'   
1987  b'Bush puts foot down on Georgian conflict'   
1986           b"Russia 'ends Georgia operation'"   

                                                   Top3  \
1988  b'Russia Today: Columns of troops roll into So...   
1987  b"Jewish Georgian minister: Thanks to Israeli ...   
1986  b'"If we had no sexual harassment we would hav...   

                                                   Top4  \
1988  b'Russian tanks are moving towards the capital...   
1987  b'Georgian army flees in disarray as Russians ...   
1986  b"Al-Qa'eda is losing support in Iraq because ...   

                                                   Top5  \
1988  b"Afghan children raped with 'impunity,' U.N. ...   
1987      b"Olympic opening ceremony fireworks 'faked'"   
1986  b'Ceasefire in Georgia: Putin Outmaneuvers the...   

                                                   Top6  \
1988  b'150 Russian tanks have entered South Ossetia...   
1987  b'What were the Mossad with fraudulent New Zea...   
1986  b'Why Microsoft and Intel tried to kill the XO...   

                                                   Top7  \
1988  b"Breaking: Georgia invades South Ossetia, Rus...   
1987  b'Russia angered by Israeli military sale to G...   
1986  b'Stratfor: The Russo-Georgian War and the Bal...   

                                                   Top8  ...         Close  \
1988  b"The 'enemy combatent' trials are nothing but...  ...  11734.320312   
1987  b'An American citizen living in S.Ossetia blam...  ...  11782.349609   
1986  b"I'm Trying to Get a Sense of This Whole Geor...  ...  11642.469727   

         Volume     Adj Close  \
1988  212830000  11734.320312   
1987  183190000  11782.349609   
1986  173590000  11642.469727   

                                    ConcatenatedTopNews  Polarity  \
1988  Georgia downs two Russian warplanes as countri... -0.048568   
1987  Why wont America and Nato help us? If they won...  0.121956   
1986  Remember that adorable 9-year-old who sang at ... -0.044302   

     Subjectivity Compound Negative Positive Neutral  
1988     0.267549  -0.9983    0.212    0.059   0.729  
1987     0.374806  -0.9977    0.202    0.061   0.738  
1986     0.536234  -0.9975    0.225    0.091   0.684  

[3 rows x 40 columns]

### Feature Selection

In [54]:
#Keep only necessary columns from the data dataframe
columns = ["Open", "High", "Low", "Volume", "Polarity", "Subjectivity", "Compound", "Negative", "Positive", "Neutral", "Label"]
data2 = data[columns]
data2.head(3)

Open          High           Low     Volume  Polarity  \
1988  11432.089844  11759.959961  11388.040039  212830000 -0.048568   
1987  11729.669922  11867.110352  11675.530273  183190000  0.121956   
1986  11781.700195  11782.349609  11601.519531  173590000 -0.044302   

      Subjectivity  Compound  Negative  Positive  Neutral  Label  
1988      0.267549   -0.9983     0.212     0.059    0.729      0  
1987      0.374806   -0.9977     0.202     0.061    0.738      1  
1986      0.536234   -0.9975     0.225     0.091    0.684      0

In [56]:
#Creating the feature dataset
X = data2
X = np.array(X.drop(['Label'], 1))

#Creating the target dataset
y = np.array(data2['Label'])

#Split the data into training and test sets
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

### Training the Model

In [57]:
#Creating and training the model
model = LinearDiscriminantAnalysis().fit(x_train, y_train)

In [58]:
#Obtain the predictions of the model
pred = model.predict(x_test)
pred

array([1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1,
       0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0,
       1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1,
       0, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1,
       1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1,
       0, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1,
       0, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1,
       1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 0, 1,
       0, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 0,
       0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0,
       1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0,
       0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 1, 1,

In [59]:
y_test

array([1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 1,
       0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1, 0, 0, 1, 1,
       1, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0, 1,
       0, 1, 0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1, 1, 0,
       1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1,
       1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 1,
       0, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0,
       1, 1, 0, 1, 1, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 1,
       1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 1,
       0, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1,
       0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0,
       1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0,
       0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 1, 1, 1,

### Model Evaluation

In [60]:
#Model Metrics
print(classification_report(y_test, pred))

              precision    recall  f1-score   support

           0       0.87      0.79      0.83       193
           1       0.82      0.89      0.85       205

    accuracy                           0.84       398
   macro avg       0.84      0.84      0.84       398
weighted avg       0.84      0.84      0.84       398



We obtain an accuracy of 84% with this model.